File: src/main/java/mysql/hibernate/entity/GuildEntity.java
package mysql.hibernate.entity;

import constants.Language;
import core.assets.GuildAsset;
import core.cache.ServerPatreonBoostCache;
import mysql.hibernate.template.HibernateEntity;
import mysql.modules.guild.DBGuild;

import javax.persistence.Embedded;
import javax.persistence.Entity;
import javax.persistence.Id;
import java.util.Locale;

@Entity(name = "Guild")
public class GuildEntity extends HibernateEntity implements GuildAsset {

    @Id
    private String guildId;

    private String prefix = "L.";
    private String language = Language.EN.name();
    private boolean removeAuthorMessage = false;

    @Embedded
    FisheryEntity fishery = new FisheryEntity();

    public GuildEntity(String guildId) {
        this.guildId = guildId;
    }

    public GuildEntity(String guildId, String prefix, String language, boolean removeAuthorMessage) {
        this.guildId = guildId;
        this.prefix = prefix;
        this.language = language;
        this.removeAuthorMessage = removeAuthorMessage;
    }

    public GuildEntity() {
    }

    @Override
    public long getGuildId() {
        return Long.parseLong(guildId);
    }

    public String getPrefix() {
        return prefix;
    }

    public void setPrefix(String prefix) {
        this.prefix = prefix;
        DBGuild.getInstance().retrieve(Long.parseLong(guildId)).setPrefix(prefix); // TODO: remove after migration
    }

    public Language getLanguage() {
        return Language.valueOf(language);
    }

    public void setLanguage(Language language) {
        this.language = language.name();
        DBGuild.getInstance().retrieve(Long.parseLong(guildId)).setLocale(language.getLocale()); // TODO: remove after migration
    }

    public Locale getLocale() {
        return getLanguage().getLocale();
    }

    public boolean getRemoveAuthorMessage() {
        return removeAuthorMessage;
    }

    public boolean getRemoveAuthorMessageEffectively() {
        return getRemoveAuthorMessage() && ServerPatreonBoostCache.get(Long.parseLong(guildId));
    }

    public void setRemoveAuthorMessage(boolean removeAuthorMessage) {
        this.removeAuthorMessage = removeAuthorMessage;
        DBGuild.getInstance().retrieve(Long.parseLong(guildId)).setCommandAuthorMessageRemove(removeAuthorMessage); // TODO: remove after migration
    }

    public FisheryEntity getFishery() {
        fishery.setHibernateEntity(this);
        return fishery;
    }
}

File: src/main/java/mysql/hibernate/entity/GuildEntity.kt
package mysql.hibernate.entity

import constants.Language
import core.assets.GuildAsset
import core.cache.ServerPatreonBoostCache
import mysql.hibernate.template.HibernateEntity
import org.hibernate.annotations.SortNatural
import java.util.*
import javax.persistence.*


@Entity(name = "Guild")
class GuildEntity(key: String) : HibernateEntity(), GuildAsset {

    @Id
    private val guildId = key

    @Column(name = "prefix")
    private var _prefix: String? = null
    var prefix: String
        get() = _prefix ?: "L."
        set(value) {
            _prefix = value
        }

    @Column(name = "language")
    @Enumerated(EnumType.STRING)
    private var _language: Language? = null
    var language: Language
        get() = _language ?: Language.EN
        set(value) {
            _language = value
        }
    val locale: Locale
        get() = language.locale

    @Column(name = "removeAuthorMessage")
    private var _removeAuthorMessage: Boolean? = null
    var removeAuthorMessage: Boolean
        get() = _removeAuthorMessage ?: false
        set(value) {
            _removeAuthorMessage = value
        }
    val removeAuthorMessageEffectively: Boolean
        get() = removeAuthorMessage && ServerPatreonBoostCache.get(guildId.toLong())

    @Embedded
    @Column(name = FISHERY)
    val fishery = FisheryEntity()

    @Embedded
    @Column(name = INVITE_FILTER)
    val inviteFilter = InviteFilterEntity()

    @Embedded
    @Column(name = WORD_FILTER)
    val wordFilter = WordFilterEntity()

    @ElementCollection
    @SortNatural
    val customCommands = sortedMapOf<String, CustomCommandEntity>()


    constructor() : this("0")

    override fun getGuildId(): Long {
        return guildId.toLong()
    }

    @PostLoad
    override fun postLoad() {
        fishery.postLoad(this)
        inviteFilter.postLoad(this)
        wordFilter.postLoad(this)
        customCommands.values.forEach { it.postLoad(this) }
    }

}
